src/typecheck.rs

```diff
@@ -1,11 +1,1384 @@
+                    if let Type::NumT(_) = lhs_type {
+                        if lhs_type != rhs_type {
+                            return Err(TypeError::TypeMismatch {
+                                expected: lhs_type,
+                                actual: rhs_type,
+                            });
+                        }
+                        Ok(lhs_type)
+                    } else {
+                        Err(TypeError::TypeMismatch {
+                            expected: Type::NumT(0..1),
+                            actual: lhs_type,
+                        })
+                    }
+                }
+                // Comparison operations
+                BOp::Lt | BOp::Lte => {
+                    if lhs_type != rhs_type {
+                        return Err(TypeError::TypeMismatch {
+                            expected: lhs_type,
+                            actual: rhs_type,
+                        });
+                    }
+                    if let Type::NumT(_) = lhs_type {
+                        Ok(Type::BoolT)
+                    } else {
+                        Err(TypeError::TypeMismatch {
+                            expected: Type::NumT(0..1),
+                            actual: lhs_type,
+                        })
+                    }
+                }
+                // Equality and inequality can compare symbols as well as numbers
+                BOp::Eq | BOp::Ne => {
+                    if lhs_type != rhs_type {
+                        return Err(TypeError::TypeMismatch {
+                            expected: lhs_type,
+                            actual: rhs_type,
+                        });
+                    }
+                    // Allow comparison of numbers or symbols
+                    if matches!(lhs_type, Type::NumT(_) | Type::SymT) {
+                        Ok(Type::BoolT)
+                    } else {
+                        Ok(Type::BoolT)
+                    }
+                }
+                // Logical operations are always of type BoolT
+                BOp::And | BOp::Or => {
+                    if lhs_type != rhs_type {
+                        return Err(TypeError::TypeMismatch {
+                            expected: lhs_type,
+                            actual: rhs_type,
+                        });
+                    }
+                    if lhs_type == Type::BoolT {
+                        Ok(Type::BoolT)
+                    } else {
+                        Err(TypeError::TypeMismatch {
+                            expected: Type::BoolT,
+                            actual: lhs_type,
+                        })
+                    }
+                }
+            }
+        }
+        // Unary operations are always of the type of the inner expression
+        Expr::UOp { op, inner } => {
+            let inner_type = typeck_expr(inner, ctx)?;
+            match op {
+                UOp::Not => {
```

Comment:
you can lift the conditionals to these cases and have a single error case if you give up on the `expected` field.